src/shared/tpm2-util.c

```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
-                                .scheme.scheme = TPM2_ALG_NULL,
-                                .keyBits = 2048,
                         },
                 },
         };
 
+        return &templ[flags];
+}
+
+/*
+ * Why and what is an SRK?
+ * TL;DR provides a working space for those without owner auth.
+ * The idea is that the TPM Owner might not be the computer owner/admin. I.E. the person performing
+ * cryptenroll might not have owner auth. In the current model, the owner auth has to be empty,
+ * which means that if I as a regular user want to do something, I would need to clear owner auth
+ * and re-establish it after the task is done. This is the WHY the SRK exists. The SRK exists to be
+ * a scratch space for regular users, which is why the password is emptyauth.
+ */
+#define SRK_HANDLE 0x81000001
+
+static int tpm2_get_srk(
+                Tpm2Context *c,
+                TPMI_ALG_PUBLIC alg,
+                TPMI_ALG_PUBLIC *ret_alg,
+                Tpm2Handle *ret_primary) {
+
+        TPMI_YES_NO more_data;
+        ESYS_TR primary = ESYS_TR_NONE;
+        _cleanup_(Esys_Freep) TPMS_CAPABILITY_DATA *cap_data = NULL;
+
+        assert(ret_primary);
+
+        TSS2_RC rc = sym_Esys_GetCapability(c->esys_context,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        TPM2_CAP_HANDLES,
+                        SRK_HANDLE,
+                        1,
+                        &more_data,
+                        &cap_data);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                       "Failed to enumerate handles searching for SRK: %s",
+                                       sym_Tss2_RC_Decode(rc));
+
+        /* Did Not find SRK, indicate this by returning 0 and ESYS_TR_NONE */
+        if (cap_data->data.handles.count == 0) {
+                ret_primary->esys_handle = ESYS_TR_NONE;
+
+                if (ret_alg)
+                        *ret_alg = 0;
+                return 0;
+        }
+
+        log_debug("Found SRK on TPM.");
+
+        /* convert the raw handle to an ESYS_TR */
+        TPM2_HANDLE handle = cap_data->data.handles.handle[0];
+        rc = sym_Esys_TR_FromTPMPublic(c->esys_context,
+                        handle,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Make sure the alg is what we were expecting */
+        _cleanup_(Esys_Freep) TPM2B_PUBLIC *out_public = NULL;
+        _cleanup_(Esys_Freep) TPM2B_NAME *out_name = NULL, *out_qname = NULL;
```

Comment:
why are you reading the name/qname? it's not being used at all

Reply:
Oh I guess ESAPI allows null...